python/cudf/cudf/tests/test_dataframe.py

```diff
@@ -9065,6 +9065,149 @@ def test_dataframe_add_suffix():
     assert_eq(got, expected)
 
 
+@pytest.mark.parametrize(
+    "data, gkey",
+    [
+        (
+            {
+                "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+                "val1": [5, 4, 6, 4, 8, 7, 4, 5, 2],
+                "val2": [4, 5, 6, 1, 2, 9, 8, 5, 1],
+                "val3": [4, 5, 6, 1, 2, 9, 8, 5, 1],
+            },
+            ["id", "val1", "val2"],
```

Comment:
Do some of these test cases result in covariance matrices that are larger than 2x2? I would like to have a test for a 3x3 (or larger) case but did not see one. This case would result in a 3x3 matrix if we only grouped by `"id"`, but it wouldn't be as good of a test of MultiIndex behavior. Maybe copy it and include both cases with different groupby keys?
```suggestion
            ["id"],
```

Reply:
A little confused about the last part of your comment, there's already a separate test for a multiindex dataframe?

Reply:
I believe what @bdice meant here is that the actual test run is different from your example above :). The groupby keys you provided is `["id", "val1", "val2"]` for a dataframe that has four columns` ['id', 'val1', 'val2', 'val3']`. This only gives 1 free variable, and in fact it never creates a cov matrix larger than 1x1 lol. You should reduce number of key columns as suggested above.

Good catch @bdice !

Reply:
Yes, basically. I think there are two valuable cases to be covered here:
1. A nontrivial MultiIndex, which requires grouping by several keys.
2. A nontrivial covariance matrix (3x3 or larger), which requires at least 3 columns _not_ in the groupby keys.

Reply:
> A little confused about the last part of your comment, there's already a separate test for a multiindex dataframe?
> 
> Also with your suggested change, some tests fail when `min_per =[1,3`] for `ddof=[3,4,5]` : `E AssertionError: DataFrame.iloc[:, 0] (column name="val1") are different`
> 
> An example:
> 
> ```
> In [4]: pdf
> Out[4]: 
>   id  val1  val2  val3
> 0  a     5     4     4
> 1  a     4     5     5
> 2  a     6     6     6
> 3  b     4     1     1
> 4  b     8     2     2
> 5  b     7     9     9
> 6  c     4     8     8
> 7  c     5     5     5
> 8  c     2     1     1
> 
> In [5]: gdf = cudf.from_pandas(pdf)
>    ...: gdf.groupby("id").cov(min_periods=1, ddof=4)
> Out[5]: 
>          val1  val2  val3
> id                       
> a  val1  <NA>  <NA>  <NA>
>    val2  <NA>  <NA>  <NA>
>    val3  <NA>  <NA>  <NA>
> b  val1  <NA>  <NA>  <NA>
>    val2  <NA>  <NA>  <NA>
>    val3  <NA>  <NA>  <NA>
> c  val1  <NA>  <NA>  <NA>
>    val2  <NA>  <NA>  <NA>
>    val3  <NA>  <NA>  <NA>
> 
> In [6]: pdf.groupby("id").cov(min_periods=1, ddof=4)
> /home/nfs/skirui/dev/rapids/compose/etc/conda/cuda_11.5/envs/rapids/lib/python3.8/site-packages/pandas/core/frame.py:9524: RuntimeWarning: Degrees of freedom <= 0 for slice
>   base_cov = np.cov(mat.T, ddof=ddof)
> /home/nfs/skirui/dev/rapids/compose/etc/conda/cuda_11.5/envs/rapids/lib/python3.8/site-packages/numpy/lib/function_base.py:2542: RuntimeWarning: divide by zero encountered in true_divide
>   c *= np.true_divide(1, fact)
> Out[6]: 
>          val1  val2  val3
> id                       
> a  val1   inf   inf   inf
>    val2   inf   inf   inf
>    val3   inf   inf   inf
> b  val1   inf   inf   inf
>    val2   inf   inf   inf
>    val3   inf   inf   inf
> c  val1   inf   inf   inf
>    val2   inf   inf   inf
>    val3   inf   inf   inf
>  
> ```

@isVoid  @bdice  Actually the results above is from running the `original test case with groupby_key = ["id"]`  (per Bradley's suggestion)
As you can see there's mismatch from pandas causing failures: `E   AssertionError: DataFrame.iloc[:, 0] (column name="val1") are different`
Not sure how to handle this difference  in behavior, any thoughts?

Reply:
> Yes, basically. I think there are two valuable cases to be covered here:
> 
> 1. A nontrivial MultiIndex (2 or more levels), which requires grouping by several keys.
> 2. A nontrivial covariance matrix (3x3 or larger), which requires at least 3 columns _not_ in the groupby keys.

@bdice 
I've covered point 2  with the example [above](https://github.com/rapidsai/cudf/pull/9889#discussion_r799916544)
For point 1, I meant that I already have a separate test `test_groupby_covariance_multiindex_dataframe()` that does exactly this ( refer to most recent file edits)

Reply:
It seems like the div by zero behavior from libcudf is different from pandas. The example you provided has 3 valid observations and `ddof==4`, thus div by 0. Is this considered a standard handling from libcudf?

cc @karthikeyann

Reply:
@karthikeyann  @bdice

Reply:
@skirui-source I'm not exactly sure what I was supposed to reply to on this comment thread.

Adding some detail to @isVoid's observation, here is a minimal example showing a few differences in behavior. I would regard this as a bug for not giving the same results as pandas, but it affects both `df.cov()` and `df.groupby(...).cov()` so some changes may be needed outside of this PR. (Note that this is _not_ related to the pandas bug 45814 that was filed regarding `ddof` with missing data, because no data is missing here.)
```python
import pandas as pd
import cudf

pdf = pd.DataFrame({"a": [0, 0], "b": [0, 1]})
gdf = cudf.from_pandas(pdf)
print("pandas cov")
print(pdf.cov(ddof=2))
print("cudf cov")
print(gdf.cov(ddof=2))
print("pandas groupby cov")
print(pdf.groupby("i").cov(ddof=2))
print("cudf groupby cov")
print(gdf.groupby("i").cov(ddof=2))
```

Results (warnings not shown):
```python
pandas cov                                                                                                                                                                                    
    i    a
i NaN  NaN
a NaN  inf

cudf cov
     i    a
i  0.0  0.0
a  0.0  0.5

pandas groupby cov
       a
i       
0 a  inf
cudf groupby cov
        a
i        
0 a  <NA>
```

This is particularly concerning because cudf isn't self-consistent between `df.cov()` and `df.groupby(...).cov()`.

Reply:
Link: https://github.com/rapidsai/cudf/issues/10303